File: python/2-add-two-numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        def nodeToNumber(node):
            if node.next == None:
                return node.val
            return node.val + nodeToNumber(node.next) * 10
        
        def numberToNode(number):
            if number // 10 == 0:
                return ListNode(number)
            digit = number % 10
            node = ListNode(digit)
            node.next = numberToNode(number // 10)
            return node
        
        num1 = nodeToNumber(l1)
        num2 = nodeToNumber(l2)
        return numberToNode(num1 + num2)

File: js/2-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {

    var ans = new ListNode(0)
    var head = ans
    var partial = 0
    var carry = 0

    while (l1 !== null || l2 !== null || partial > 0) {
        if (l1 !== null) {
            partial = partial + l1.val
            l1 = l1.next
        }
        if (l2 !== null) {
            partial = partial + l2.val
            l2 = l2.next
        }
        if (partial >= 10) {
            carry = 1
            partial -= 10
        }

        head.next = new ListNode(partial)
        head = head.next

        partial = carry
        carry = 0
    }

    return ans.next

};
